src/arch/arm/64/kernel/vspace.c

```diff
@@ -1183,55 +1186,43 @@ static vspace_root_t getASIDBindCB(asid_t asid)
     asid_map_t asid_map = asidPool->array[asid & MASK(asidLowBits)];
     assert(asid_map_get_type(asid_map) == asid_map_asid_map_vspace);
 
-    vspace_root_t *vtable = (vspace_root_t *)asid_map_asid_map_vspace_get_vspace_root(asid_map);;
-    assert(vtable);
-
-    return vtable[VTABLE_SMMU_SLOT];
+    return asid_map_asid_map_vspace_get_bind_cb(asid_map);
 }
 
 void increaseASIDBindCB(asid_t asid)
 {
     asid_pool_t *asidPool;
-    vspace_root_t stored_info;
 
     asidPool = armKSASIDTable[asid >> asidLowBits];
     assert(asidPool);
 
-    asid_map_t asid_map = asidPool->array[asid & MASK(asidLowBits)];
-    assert(asid_map_get_type(asid_map) == asid_map_asid_map_vspace);
+    asid_map_t *asid_map = &asidPool->array[asid & MASK(asidLowBits)];
+    assert(asid_map_ptr_get_type(asid_map) == asid_map_asid_map_vspace);
 
-    vspace_root_t *vtable = (vspace_root_t *)asid_map_asid_map_vspace_get_vspace_root(asid_map);;
-    assert(vtable);
-
-    stored_info = vtable[VTABLE_SMMU_SLOT];
-    vtable[VTABLE_SMMU_SLOT] = vtable_invalid_smmu_new(vtable_invalid_get_bind_cb(stored_info) + 1);
+    asid_map_asid_map_vspace_ptr_set_bind_cb(asid_map, asid_map_asid_map_vspace_ptr_get_bind_cb(asid_map) + 1);
```

Comment:
Good point. This is independent of this PR, but I think there is a bug in this path, yes, because the decode path does not check that the increment is safe (if assertions are turned on, this would fail a bf generator assertion I think; if they are turned off it would wrap, which is not clear that we want here).

The check should be in decode, though, not here, because in here it is too late to do anything about it. It's actually unclear to me how this is supposed to work when the CB space is used up.

We should create an issue for this.

Reply:
There is a maximum number of CBs allowed in an SMMU implementation defined by the spec to be 128 and CBs can only be assigned to a single VSpace. 8 bits was chosen to be large enough to support every CB being assigned to the same VSpace on any SMMUv2 implementation so it shouldn't be able to overflow due to existing checks in `decodeARMCBInvocation` and `decodeARMCBControlInvocation`.

Reply:
In any case it is a separate issue, happy to leave it for another PR or an issue (still need to double check that decode path -- the first time I looked I didn't see a check that would constrain the overflow)

Reply:
It's constrained by the upper bound on number of CBs that can exist in the system which is the number of slots in the static CB table: `extern bool_t smmuStateCBTable[SMMU_MAX_CB];`  SMMU_MAX_CB cannot be greater than 128.